README.md
# tp-stocks

 * First, [install stack](https://docs.haskellstack.org/en/stable/install_and_upgrade/)
 * Recommended : install [VSCode](https://code.visualstudio.com/download), with the *Haskell* extension
 * Run `./test.sh`
 * Edit the `SampleStudent.hs` file

Do not edit any other file!

You will be graded on the `SampleStudent.hs` file only. Do not send me any other file!


ChangeLog.md
# Changelog for tp-stocks

## Unreleased changes


test/Spec.hs
module Main where

import Basics as B
import qualified Inventory as I
import InventorySpec (checkoutTests)
import qualified S2021.SampleStudent as S
import Test.Hspec (SpecWith, describe, hspec, it, shouldBe, xdescribe)
import Test.Hspec.QuickCheck (prop)

basics :: B.Solution -> SpecWith ()
basics sol = describe "basics" $ do
  describe "simple operations" $ do
    prop "add" (\a b -> _add sol a b == a + b)
    prop "divide" (\a b -> b == 0 || _divide sol a b == a `div` b)
    prop "divide' /= 0" (\a b -> b == 0 || _divide' sol a b == IntSuccess (div a b))
    prop "divide' == 0" (\a -> _divide' sol a 0 == IntFailure)
    prop "divide'' /= 0" (\a b -> b == 0 || _divide'' sol a b == B.Success (div a b))
    prop "divide'' == 0" (\a -> _divide'' sol a 0 == B.Failure)
  describe "eval" $ do
    prop "Value" (\a -> _eval sol (Value a) == a)
    prop "Add" (\a b -> _eval sol (Add (Value a) (Value b)) == a + b)
    prop "Sub" (\a b -> _eval sol (Sub (Value a) (Value b)) == a - b)
    prop "Mul" (\a b -> _eval sol (Mul (Value a) (Value b)) == a * b)
    it "test 1" (_eval sol (Add (Mul (Value 5) (Value 3)) (Value 8)) `shouldBe` 5 * 3 + 8)
  describe "eval'" $ do
    prop "Value" (\a -> _eval' sol (Value a) == B.Success a)
    prop "Add" (\a b -> _eval' sol (Add (Value a) (Value b)) == B.Success (a + b))
    prop "Sub" (\a b -> _eval' sol (Sub (Value a) (Value b)) == B.Success (a - b))
    prop "Mul" (\a b -> _eval' sol (Mul (Value a) (Value b)) == B.Success (a * b))
    it "test 1" (_eval' sol (Add (Mul (Value 5) (Value 3)) (Value 8)) `shouldBe` B.Success (5 * 3 + 8))
    it "test 2" (_eval' sol (Div (Div (Value 5) (Value 3)) (Value 0)) `shouldBe` B.Failure)
    it "test 3" (_eval' sol (Div (Div (Value 5) (Value 0)) (Value 1)) `shouldBe` B.Failure)
  describe "lists" $ do
    let foo :: Int -> Maybe Char
        foo x
          | x < 0 = Just 'a'
          | x > 0 = Just 'b'
          | otherwise = Nothing
    it "listHead1" (_listHead sol (Cons 1 (Cons 2 Empty)) `shouldBe` Success (1 :: Int))
    it "listHead2" (_listHead sol (Empty :: List Int) `shouldBe` Failure)
    prop "foo" (foo 3 == Just 'b')
    prop "listSum" (\l -> _listSum sol (fromList l) == sum l)
    prop "listEq" (\l1 l2 -> _listEq sol (fromList l1) (fromList l2) == ((l1 :: [Int]) == l2))
    prop "toList" (\l -> _toList sol (fromList l) == (l :: [Int]))
    prop "lmap" (\l -> _toList sol (_lmap sol foo (fromList l)) == fmap foo l)

    xdescribe "pending tests (edit test if you wrote the ltraverse function)" $
      prop "ltraverse" (\l -> fmap (_toList sol) (_ltraverse sol foo (fromList l)) == traverse foo (l :: [Int]))

fromList :: [a] -> List a
fromList = foldr Cons Empty

inventoryspecs :: I.Solution -> SpecWith ()
inventoryspecs sol = describe "Inventory" $ do
  checkoutTests sol

main :: IO ()
main = hspec $ do
  basics S.bsolution
  inventoryspecs S.isolution


test/InventorySpec.hs
module InventorySpec (checkoutTests) where

import qualified Data.Map.Strict as M
import qualified Inventory as I
import Test.Hspec

checkoutTests :: I.Solution -> Spec
checkoutTests sol = do
  describe "getters" $ do
    let sampledesc = I.ItemDesc "name" 5 I.Book
    it "getName" $ I._getName sol sampledesc `shouldBe` "name"
    it "getPrice" $ I._getPrice sol sampledesc `shouldBe` 5
    it "getType" $ I._getType sol sampledesc `shouldBe` I.Book
  describe "getPrice" $ do
    it "known item 1" $ I._getItemPrice sol I.sampleCatalog 1 `shouldBe` Right 785
    it "known item 2" $ I._getItemPrice sol I.sampleCatalog 2 `shouldBe` Right 200
    it "known item 3" $ I._getItemPrice sol I.sampleCatalog 3 `shouldBe` Right 24700
    it "unknown item" $ I._getItemPrice sol I.sampleCatalog 10 `shouldBe` Left I.DescriptionNotFound
  describe "updateInventory" $ do
    it "empty inventory" $ I._updateInventory sol mempty (M.fromList [(1, 4)]) `shouldBe` Left I.OutOfStock
    it "empty basket" $ I._updateInventory sol (M.fromList [(1, 4)]) mempty `shouldBe` Right (M.fromList [(1, 4)])
    it "oos1" $ I._updateInventory sol (M.fromList [(1, 4), (2, 5)]) (M.fromList [(1, 5)]) `shouldBe` Left I.OutOfStock
    it "oos2" $ I._updateInventory sol (M.fromList [(1, 4), (2, 5)]) (M.fromList [(3, 1)]) `shouldBe` Left I.OutOfStock
    it "oos3" $ I._updateInventory sol (M.fromList [(1, 4), (2, 5)]) (M.fromList [(1, 5), (2, 6)]) `shouldBe` Left I.OutOfStock
    it "oos4" $ I._updateInventory sol (M.fromList [(1, 4), (2, 5)]) (M.fromList [(1, 3), (2, 6)]) `shouldBe` Left I.OutOfStock
    it "oos5" $ I._updateInventory sol (M.fromList [(1, 4), (2, 5)]) (M.fromList [(1, 5), (2, 3)]) `shouldBe` Left I.OutOfStock
    it "checkout ok" $ I._updateInventory sol (M.fromList [(1, 4), (2, 5)]) (M.fromList [(1, 2), (2, 3)]) `shouldBe` Right (M.fromList [(1, 2), (2, 2)])
    it "take all of an item" $ I._updateInventory sol (M.fromList [(1, 4), (2, 5)]) (M.fromList [(1, 4), (2, 3)]) `shouldBe` Right (M.fromList [(2, 2)])
  describe "annotatePrice" $ do
    it "empty basket" $ I._annotatePrice sol I.sampleCatalog mempty `shouldBe` Right mempty
    it "known prices" $ I._annotatePrice sol I.sampleCatalog (M.fromList [(1, 5), (2, 8)]) `shouldBe` Right (M.fromList [(1, (5, 785)), (2, (8, 200))])
    it "unknown price" $ I._annotatePrice sol I.sampleCatalog (M.fromList [(1, 5), (8, 8)]) `shouldBe` Left I.DescriptionNotFound
  describe "computePrice" $ do
    it "no rebates" $ I._computePrice sol [] (M.fromList [(1, (5, 8)), (56, (12, 32))]) `shouldBe` (5 * 8 + 12 * 32)
  describe "checkout" $ do
    describe "no rebates" $ do
      it "empty basket" $
        I._checkout sol I.sampleCatalog [] I.sampleInventory mempty `shouldBe` Right (I.sampleInventory, 0)
      it "out of stock" $
        I._checkout sol I.sampleCatalog [] I.sampleInventory (M.singleton 2 5)
          `shouldBe` Left I.OutOfStock
      it "buy all items" $
        I._checkout sol I.sampleCatalog [] I.sampleInventory (M.singleton 2 4)
          `shouldBe` Right (M.delete 2 I.sampleInventory, 4 * 200)
      it "something that has no description, but is in stock" $
        I._checkout sol I.sampleCatalog [] (M.insert 12 1 I.sampleInventory) (M.singleton 12 1)
          `shouldBe` Left I.DescriptionNotFound
      it "buy it all" $
        I._checkout sol I.sampleCatalog [] I.sampleInventory I.sampleInventory
          `shouldBe` Right (mempty, 12 * 785 + 4 * 200 + 8 * 24700 + 1 * 19700 + 43234 * 1500)
    describe "basic rebate" $ do
      let basket = M.fromList [(1, 3), (4, 1)]
          remaining =
            M.fromList
              [ (1, 9),
                (2, 4),
                (3, 8),
                (5, 43234)
              ]
      it "one item on sale" $
        I._checkout sol I.sampleCatalog [I.Rebate 1 700] I.sampleInventory basket
          `shouldBe` Right (remaining, 3 * 700 + 1 * 19700)
      it "two items on sale" $
        I._checkout sol I.sampleCatalog [I.Rebate 1 700, I.Rebate 4 5000] I.sampleInventory basket
          `shouldBe` Right (remaining, 3 * 700 + 1 * 5000)
      it "bad sale" $
        I._checkout sol I.sampleCatalog [I.Rebate 1 800] I.sampleInventory basket
          `shouldBe` Right (remaining, 3 * 785 + 1 * 19700)
      it "one good, and one bad sale" $
        I._checkout sol I.sampleCatalog [I.Rebate 1 700, I.Rebate 4 50000] I.sampleInventory basket
          `shouldBe` Right (remaining, 3 * 700 + 1 * 19700)
    describe "grouped rebate" $ do
      let basket = M.fromList [(1, 8)]
          remaining =
            M.fromList
              [ (1, 4),
                (2, 4),
                (3, 8),
                (4, 1),
                (5, 43234)
              ]
      it "simple grouped rebate" $
        I._checkout sol I.sampleCatalog [I.Grouped 1 8 1000] I.sampleInventory basket
          `shouldBe` Right (remaining, 1000)
      it "bad grouped rebate" $
        I._checkout sol I.sampleCatalog [I.Grouped 1 8 100000] I.sampleInventory basket
          `shouldBe` Right (remaining, 8 * 785)
      it "multi rebate 1" $
        I._checkout sol I.sampleCatalog [I.Grouped 1 4 1000] I.sampleInventory basket
          `shouldBe` Right (remaining, 2000)
      it "multi rebate 2" $
        I._checkout sol I.sampleCatalog [I.Grouped 1 5 3500, I.Grouped 1 3 2000] I.sampleInventory basket
          `shouldBe` Right (remaining, 3500 + 2000)
      it "mixed rebates" $
        I._checkout sol I.sampleCatalog [I.Grouped 1 5 6500, I.Grouped 1 3 2000] I.sampleInventory basket
          `shouldBe` Right (remaining, 2 * 785 + 4000)
    describe "nth free" $ do
      let basket = M.fromList [(1, 8)]
          remaining =
            M.fromList
              [ (1, 4),
                (2, 4),
                (3, 8),
                (4, 1),
                (5, 43234)
              ]
      it "simple nth" $
        I._checkout sol I.sampleCatalog [I.NthFree 1 7] I.sampleInventory basket
          `shouldBe` Right (remaining, 7 * 785)
    describe "LEVEL UP" $ do
      let basket = M.fromList [(1, 8)]
          remaining =
            M.fromList
              [ (1, 4),
                (2, 4),
                (3, 8),
                (4, 1),
                (5, 43234)
              ]
      it "bad" $
        I._checkout sol I.sampleCatalog [I.Grouped (-1) 6 1000, I.Grouped 133 2 500] I.sampleInventory basket
          `shouldBe` Right (remaining, 8 * 785)
      it "same product id, grouped" $
        I._checkout sol I.sampleCatalog [I.Grouped 1 6 1000, I.Grouped 1 2 500] I.sampleInventory basket
          `shouldBe` Right (remaining, 1500)
      it "same product id, mixed" $
        I._checkout sol I.sampleCatalog [I.Grouped 1 5 1000, I.NthFree 1 2] I.sampleInventory basket
          `shouldBe` Right (remaining, 1000 + 2 * 785)
      it "same product id, mixed, useless" $
        I._checkout sol I.sampleCatalog [I.Grouped 1 6 1000, I.NthFree 1 2] I.sampleInventory basket
          `shouldBe` Right (remaining, 1000 + 2 * 785)
      it "same product, size X" $
        I._checkout sol I.sampleCatalog (map (I.Rebate 1 . I.Price) [500 .. 520]) I.sampleInventory basket
          `shouldBe` Right (remaining, 500 * 8)
      it "same product, size XL" $ do
        -- pendingWith "uncomment this if you feel confident"
        I._checkout sol I.sampleCatalog (map (I.Rebate 1 . I.Price) (reverse [500 .. 600])) I.sampleInventory basket
          `shouldBe` Right (remaining, 500 * 8)
      it "same product, size XXL" $ do
        -- pendingWith "uncomment this if you feel confident"
        I._checkout sol I.sampleCatalog (map (I.Rebate 1 . I.Price) [500 .. 1000]) I.sampleInventory basket
          `shouldBe` Right (remaining, 500 * 8)


src/Basics.hs
{-# LANGUAGE RankNTypes #-}

module Basics where

-- This is a sum type. A value of type "IntResult" can either be
-- a IntSuccess containing a Int, or a IntFailure.
--
-- This will be used to identify operations that fail (for example, because
-- of a divide by 0 operation).
data IntResult
  = IntSuccess Int
  | IntFailure
  deriving (Show, Eq)

-- | This is a generic (parametric) type. It represents something that can
-- either fail, or succeed with a value of type "a".
data Result a
  = Success a
  | Failure
  deriving (Show, Eq)

-- | Represent the kind of expressions one can compute on a pocket calculator.
data Operation
  = Add Operation Operation
  | Sub Operation Operation
  | Mul Operation Operation
  | Div Operation Operation
  | Value Int
  deriving (Show, Eq)

-- | A (linked) list is either empty, or contains an element (head) and a list (tail).
data List a
  = Empty
  | Cons a (List a)
  deriving (Show)

data Solution = Solution
  { _add :: Int -> Int -> Int,
    _divide :: Int -> Int -> Int,
    _divide' :: Int -> Int -> IntResult,
    _divide'' :: Int -> Int -> Result Int,
    _eval :: Operation -> Int,
    _eval' :: Operation -> Result Int,
    _listHead :: forall a. List a -> Result a,
    _listTail :: forall a. List a -> Result (List a),
    _listSum :: List Int -> Int,
    _listEq :: forall a. Eq a => List a -> List a -> Bool,
    _toList :: forall a. List a -> [a],
    _lmap :: forall a b. (a -> b) -> List a -> List b,
    _ltraverse :: forall f a b. Applicative f => (a -> f b) -> List a -> f (List b)
  }

src/S2021/SampleStudent.hs
module S2021.SampleStudent where

import Basics as B
import qualified Data.Map.Strict as M
import Inventory as I

students :: [String]
students = ["Student 1", "Student 2"]

-- PART 1 - basics

-- | Write a functions that adds two Int.
add :: Int -> Int -> Int
add = undefined

-- | Write a functions that divides two Int.
divide :: Int -> Int -> Int
divide = undefined

-- | Rewrite a function that divides two Ints, but that explicitely
-- fails when it divides by 0, or explicitely succeeds.
divide' :: Int -> Int -> IntResult
divide' = undefined

-- | Write this function again, but this time using the generic type.
divide'' :: Int -> Int -> Result Int
divide'' = undefined

-- | This is (5 * 3.4) + 8
-- run it in GHCi
example1 :: Int
example1 = eval (Add (Mul (Value 5) (Value 3)) (Value 8))

-- | What is this ?
example2 :: Int
example2 = eval (Div (Mul (Value 5) (Value 3)) (Value 0))

-- | From an operation, compute its value.
eval :: Operation -> Int
eval = undefined

-- | Try it in the REPL with example2!
-- Now write the correct function:
eval' :: Operation -> Result Int
eval' = undefined

-- | Examples to run in GHCi
example1' :: Result Int
example1' = eval' (Add (Mul (Value 5) (Value 3)) (Value 8))

example2' :: Result Int
example2' = eval' (Div (Mul (Value 5) (Value 3)) (Value 0))

-- PART 2 - lists

-- | Returns the first element of a list.
listHead :: List a -> Result a
listHead = undefined

head1 :: Result Int
head1 = listHead Empty -- Failure

head2 :: Result Int
head2 = listHead (Cons 5 (Cons 4 Empty)) -- success 5

-- | Returns the tail of a list.
listTail :: List a -> Result (List a)
listTail = undefined

tail1 :: Result (List Int)
tail1 = listTail Empty -- Failure

tail2 :: Result (List Int)
tail2 = listTail (Cons 5 (Cons 4 Empty)) -- success (Cons 4 Empty)

-- | Sum of all integers in a list.
listSum :: List Int -> Int
listSum = undefined

-- | Compare two lists for equality.
listEq :: Eq a => List a -> List a -> Bool
listEq = undefined

-- | Converts our list type into Haskell's built-in list type.
toList :: List a -> [a]
toList = undefined

-- Given a function, converts all elements of a list.
lmap ::
  (a -> b) ->
  List a ->
  List b
lmap = undefined

-- optional assignment
-- Uncomment the relevant test if you wrote it!
ltraverse ::
  Applicative f =>
  (a -> f b) ->
  List a ->
  f (List b)
ltraverse = undefined

-- PART 3 - homework

-- | Returns the name of an item description.
--
-- You need to use pattern matching.
getName :: ItemDesc -> String
getName = undefined

-- | Returns the price of an item description.
--
-- You need to use pattern matching.
getPrice :: ItemDesc -> Price
getPrice = undefined

-- | Returns the type of an item description.
--
-- You need to use pattern matching.
getType :: ItemDesc -> ItemType
getType = undefined

-- | Retrieves an item price from the catalog. Can fail if the item is
-- unknown.
getItemPrice ::
  Catalog ->
  ItemId ->
  Either Problem Price
getItemPrice = undefined

-- | Update the inventory by removing all items from the basket. This could
-- fail with `Left OutOfStock` if the inventory is not sufficiently
-- stocked.
updateInventory ::
  Inventory ->
  Basket ->
  Either Problem Inventory
updateInventory = undefined

-- | Given a catalog, and a basket, annotate prices of the basket
-- can fail if the description is not found!
annotatePrice :: Catalog -> Basket -> Either Problem (M.Map ItemId (Quantity, Price))
annotatePrice = undefined

-- | Given a list of rebates, and a shopping basket with prices attached,
-- return the total cost of the basket.
-- 
-- extra: take into account the rebates to find the best price possible
computePrice ::
  [Rebates] ->
  M.Map ItemId (Quantity, Price) ->
  Price
computePrice = undefined

-- | The checkout functions computes how much a customer should pay for
-- a given basket, and updates the inventory. It can also fail!
--
-- It is recommended to first fill the next functions, as they will be
-- quite useful for writing this one!
checkout ::
  -- | products catalog
  Catalog ->
  -- | currently available rebates
  [Rebates] ->
  -- | current inventory
  Inventory ->
  -- | customer basket
  Basket ->
  -- | if this succeeds, return the basket total cost and the updated inventory.
  Either Problem (Inventory, Price)
checkout = undefined

-- DO NOT EDIT THESE DEFINITIONS
isolution :: I.Solution
isolution = I.Solution getItemPrice updateInventory annotatePrice computePrice checkout getName getPrice getType

bsolution :: B.Solution
bsolution = B.Solution add divide divide' divide'' eval eval' listHead listTail listSum listEq toList lmap ltraverse


src/Inventory.hs
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

module Inventory where

--
--   DO NOT EDIT THIS FILE
--

import qualified Data.Map.Strict as M

data ItemType
  = Clothing
  | HighTech
  | Book
  deriving (Show, Eq, Ord, Enum, Bounded)

newtype ItemId = ItemId Int
  deriving (Show, Eq, Ord, Num)

newtype Quantity = Quantity Int
  deriving (Show, Eq, Ord, Num)

newtype Price = Price Integer
  deriving (Show, Eq, Ord, Num)

data ItemDesc
  = ItemDesc
      String -- name
      Price -- cost
      ItemType -- item type
  deriving (Show, Eq)

-- | A catalog maps item identifiers (key) to item descriptions (values).
type Catalog = M.Map ItemId ItemDesc

-- | An inventory maps item identifiers to quantities.
-- Absent entries (when the map does not hold the item identifier) mean
-- there is no inventory.
type Inventory = M.Map ItemId Quantity

-- | The shopping basket has the same type as the inventory.
type Basket = M.Map ItemId Quantity

sampleCatalog :: Catalog
sampleCatalog =
  M.fromList
    [ (1, ItemDesc "Neuromancer" 785 Book),
      (2, ItemDesc "Rashômon" 200 Book),
      (3, ItemDesc "Huawei P20 Lite" 24700 HighTech),
      (4, ItemDesc "Xiaomi Redmi Note 5" 19700 HighTech),
      (5, ItemDesc "WearAll - Femmes Tricoté Longue Manche Renne Noël 3D Chandail Dames Flocon De Neige Cavalier" 1500 Clothing)
    ]

sampleInventory :: Inventory
sampleInventory =
  M.fromList
    [ (1, 12),
      (2, 4),
      (3, 8),
      (4, 1),
      (5, 43234)
    ]

-- | Rebates! There are three types of rebates:
data Rebates
  = -- | The item price is modified.
    --   For example, if item 3 costs 85€, the rebate will be:
    --   > Rebate 3 85
    Rebate ItemId Price
  | -- | Item bundles with a discount price.
    -- For example, with the item #5, if buying a bundle of 3 costs 150€,
    -- the rebate will be:
    -- > Grouped 5 3 150
    Grouped ItemId Quantity Price
  | -- | If you buy N items, the next is free.
    -- For example, for item #5, if the 4th is free (meaning you get
    -- 4 items for the price of 3), the rebase will be:
    -- > NthFree 5 3
    NthFree ItemId Quantity
  deriving (Show)

-- Rebates are not automatically applied. The combination that is the
-- most advantageous to the customer should be selected among all
-- possibilities!

-- | Problems that could occur when running the checkout procedure
data Problem
  = OutOfStock
  | DescriptionNotFound
  deriving (Show, Eq)

-- | Use this function to compute the price. You should not do it yourself
-- as, if time allows, we will refactor the whole thing later with proper
-- types.
cost :: Quantity -> Price -> Price
cost (Quantity q) (Price p) = Price (fromIntegral q * p)

data Solution = Solution
  { _getItemPrice :: Catalog -> ItemId -> Either Problem Price,
    _updateInventory :: Inventory -> Basket -> Either Problem Inventory,
    _annotatePrice :: Catalog -> Basket -> Either Problem (M.Map ItemId (Quantity, Price)),
    _computePrice :: [Rebates] -> M.Map ItemId (Quantity, Price) -> Price,
    _checkout :: Catalog -> [Rebates] -> Inventory -> Basket -> Either Problem (Inventory, Price),
    _getName :: ItemDesc -> String,
    _getPrice :: ItemDesc -> Price,
    _getType :: ItemDesc -> ItemType
  }
